### PR TITLE
Add the example of Wasserstein GAN

### PR DESCRIPTION
**Pull request by bobchennan**
The major code is consistent with the ACGAN example for mnist.

I mainly change it to the Wasserstein GAN and add a progbar indicator to show the current loss value.

### CODE REVIEW

File: examples/mnist_wgan.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+from keras.backend.common import _EPSILON
+from keras.utils.generic_utils import Progbar
+import numpy as np
+
+np.random.seed(1331)
+
+K.set_image_dim_ordering('th')
+
+
+def modified_binary_crossentropy(target, output):
```
Comment (the-moliver) [inline examples/mnist_wgan.py]: I think you should just call this `wasserstein_loss` since there is really no relationship to binary cross-entropy any longer.

File: examples/mnist_wgan.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    latent_size = 100
+
+    # Adam parameters suggested in https://arxiv.org/abs/1511.06434
+    adam_lr = 0.0002
+    adam_beta_1 = 0.5
+
+    # build the discriminator
+    discriminator = build_discriminator()
+    discriminator.compile(
+        optimizer=SGD(clipvalue=0.01),
```
Comment (the-moliver) [inline examples/mnist_wgan.py]: Why are you using 3 different optimizers?

File: examples/mnist_wgan.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        loss=[modified_binary_crossentropy, 'sparse_categorical_crossentropy']
+    )
+
+    # build the generator
+    generator = build_generator(latent_size)
+    generator.compile(optimizer=Adam(lr=adam_lr, beta_1=adam_beta_1),
```
Comment (the-moliver) [inline examples/mnist_wgan.py]: Why are you using 3 different optimizers?

File: examples/mnist_wgan.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+
+    # get a fake image
+    fake = generator([latent, image_class])
+
+    # we only want to be able to train generation for the combined model
+    discriminator.trainable = False
+    fake, aux = discriminator(fake)
+    combined = Model(input=[latent, image_class], output=[fake, aux])
+
+    combined.compile(
```
Comment (the-moliver) [inline examples/mnist_wgan.py]: Why are you using 3 different optimizers?

File: examples/mnist_wgan.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+    X_test = (X_test.astype(np.float32) - 127.5) / 127.5
+    X_test = np.expand_dims(X_test, axis=1)
+
+    nb_train, nb_test = X_train.shape[0], X_test.shape[0]
+
+    train_history = defaultdict(list)
+    test_history = defaultdict(list)
+
+    for epoch in range(nb_epochs):
```
Comment (the-moliver) [inline examples/mnist_wgan.py]: You don't appear to be using the training schedule recommended from the paper. Maybe it would be a good idea for a canonical example like this unless there's a good reason not to.

File: examples/mnist_wgan.py (hunk 2 above)
Comment (bobchennan) [inline examples/mnist_wgan.py]: I will try to use the same optimizer. Since for the origin GAN it is recommended to use weaker optimizer for discriminator, I didn't change it.

File: examples/mnist_wgan.py (hunk 4 above)
Comment (bobchennan) [inline examples/mnist_wgan.py]: For discriminator we need to limit the gradient or the absolute value so it makes difference. I think two optimizers are needed.

File: examples/mnist_wgan.py (hunk 5 above)
Comment (bobchennan) [inline examples/mnist_wgan.py]: Could you please give me more explanation about the difference?

File: examples/mnist_wgan.py (hunk 2 above)
Comment (the-moliver) [inline examples/mnist_wgan.py]: In the WGAN paper, they recommend using RMSprop for everything, and clipping the weight values of the discriminator. You are using gradient clipping which is different. You should use a constraint instead.

File: examples/mnist_wgan.py (hunk 5 above)
Comment (the-moliver) [inline examples/mnist_wgan.py]: Check out the training schedule used here: https://github.com/martinarjovsky/WassersteinGAN/blob/master/main.py
Basically, you want to train the discriminator more initially, and then train it a lot every so often so that it maintains good performance.

File: examples/mnist_wgan.py (hunk 2 above)
Comment (bobchennan) [inline examples/mnist_wgan.py]: Yes but this implementation in Keras is a little bit tricky since there are no such constraint. 
One thing we can do is get all parameters and clip them.

Instead I think if you limited the gradients which is corresponding to the updates, you can also limit the range of the parameters for discriminator.

Comment (the-moliver) [inline examples/mnist_wgan.py]: No it doesn't work like that. Even if you clip the gradients the parameters can grow without bound given enough updates. You can make a custom constraint if needed. Actually I'd recommend creating this constraint as a new PR and then using it here.

File: examples/mnist_wgan.py (hunk 5 above)
Comment (bobchennan) [inline examples/mnist_wgan.py]: I see. I read that before but since MNIST is a small dataset I think 1 iteration may be enough for discriminator.

I will try to use more iterations.

File: examples/mnist_wgan.py (hunk 2 above)
Comment (bobchennan) [inline examples/mnist_wgan.py]: Yeah I will try to submit a new PR for keras. But for this constrain is there any other cases that people will use it?

Comment (the-moliver) [inline examples/mnist_wgan.py]: You can submit here instead for the time being

Comment (the-moliver) [inline examples/mnist_wgan.py]: I created a new clip constraint in keras-contrib/constraints.py you can use.

**Comment by tboquet** [comment]
@bobchennan is this PR still an active project?

**Comment by bobchennan** [comment]
Sorry but recently I am attending a conference. The result on MNIST is not outstanding and the training process is slow (for example you need to train discriminator 100 times in the first several iterations). I am still looking for better implementation.

**Comment by tboquet** [comment]
Ok, if I have time I'll try to work on this. We can wait for better results, do you know when you will be able to improve the PR?

**Comment by MarkSeygan** [comment]
I thought that the discriminator weights should be clamped. Maybe I am mistaken but gradient clipping doesn't constraint weights. Thanks for enlightening :)

**Comment by the-moliver** [comment]
I'm going to close this for now in favor of the Improved WGAN example.
